### PR TITLE
[@types/hapi] Add new point of extension (ServerMethods interface)

### PR DESCRIPTION
**Pull request by Ginden**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: - N/A, it's just point of extension for end users https://hapi.dev/api/?v=18.4.0#-servermethods

### CODE REVIEW

**Comment by saboya** [comment]
What the PR template means by this:
```Provide a URL to documentation or source code which provides context for the suggested changes```

Is for you to link the documentation that backs up the definition that you are proposing. In this case, it would be https://hapi.dev/api/?v=18.4.0#-servermethods

**Comment by SimonSchick** [comment]
You are editing the wrong package, `hapi` is no longer maintained, use `@hapi/hapi`

**Comment by Ginden** [comment]
@SimonSchick Fixed.

**Comment by SimonSchick** [comment]
Also can you explain the motivation behind this change?

**Comment by Ginden** [comment]
@SimonSchick 
I, as user of Hapi, want to add server method, eg. `add`.

So I write code like:
```
server.method('add', (a: number, b: number) => (a + b));
server.methods.add(1, 2);    // 3. Type? any
server.methods.add({foo: 'bar'}, ['foobar']); //   '[object Object]foobar'. Type? any
```

After merging this change, I can add type safety to these methods.

```
declare module '@hapi/hapi' {
   interface ServerMethods {
       add(a: number, b: number): number;
   }
}
...
server.method('add', (a: number, b: number) => (a + b));
server.methods.add(1, 2);    // 3. Type? number
server.methods.add({foo: 'bar'}, ['foobar']); //  forbidden
```

**Comment by lenovouser** [comment]
> I, as user of Hapi, want to add server method, eg. `add`.
> 
> So I write code like:
> 
> ```
> server.method('add', (a: number, b: number) => (a + b));
> server.methods.add(1, 2);    // 3. Type? any
> server.methods.add({foo: 'bar'}, ['foobar']); //   '[object Object]foobar'. Type? any
> ```
> 
> After merging this change, I can add type safety to these methods.
> 
> ```
> declare module '@hapi/hapi' {
>    interface ServerMethods {
>        add(a: number, b: number): number;
>    }
> }
> ...
> server.method('add', (a: number, b: number) => (a + b));
> server.methods.add(1, 2);    // 3. Type? number
> server.methods.add({foo: 'bar'}, ['foobar']); //  forbidden
> ```

@Ginden maybe add tests for this

**Comment by AJamesPhillips** [comment]
Ditto @lenovouser 's suggestion.  And try adding 2 methods to check it behaves as you'd want.  I (carefully) use manual typing in my own projects first to ensure the typing was what I was expecting.

Also you've got `readonly methods: ServerMethods & Util.Dictionary<ServerMethod>;` where `ServerMethods` interface implements `Util.Dictionary<ServerMethod>`... so could we not just use `readonly methods: ServerMethods;`?

**Comment by Ginden** [comment]
> so could we not just use readonly methods: ServerMethods;?

Done!

File: types/hapi/v16/index.d.ts [hunk 1]
```diff
@@ -101,7 +105,7 @@ export class Server extends Podium {
      * An object providing access to the server methods cs://hapijs.com/api/16.1.1#servermethodname-method-options} where each server method name is an object property.
      * [See docs](https://hapijs.com/api/16.1.1#servermethods)
      */
-    methods: Dictionary<ServerMethod>;
+    methods: ServerMethods & Dictionary<ServerMethod>;
```
Comment (andrewbranch) [inline types/hapi/v16/index.d.ts]: It seems like this intersection is redundant, and the other versions you edited just use `ServerMethods`—is this intentional?

**Review by andrewbranch — CHANGES_REQUESTED** [review]
This looks like a lovely change, but please add some tests as other reviewers have suggested. (IMO it’s fine to add them just to the latest version of @hapi/hapi.) Thank you!

File: types/hapi/v16/index.d.ts (hunk 1 above)
Comment (Ginden) [reply types/hapi/v16/index.d.ts]: Mistake, fixed.

**Comment by Ginden** [comment]
@andrewbranch Fixed your remarks.

**Review by andrewbranch — APPROVED** [review]

**Pull request merged**